`ct/l1`: forward progress on compaction `sink` and `source` [CT - Compaction #3]

WillemKauf opened the pull request:
This PR makes more forward progress on the compaction `sink` and `source` implementations and has a pretty fun new fixture test to demonstrate an interaction between a compaction `sink`, `source`, `reducer`, `metastore`, and `committer` all working in tandem to produce updates to L1 data committed to the `metastore`.

Notably changes are the addition of the `L1` reader in the `sink`/`source`, as well as the switch from an in memory buffer to a staging file for the compaction output.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--450c1cc6-5059-4a67-a7bc-f51c528b43b0-->
<!--TEST RESULTS-->
<!--3822a273-5b92-4e24-82c6-43e3f232d0ac-->
<!--TEST RESULTS-->
<!--5ebd683c-1ccb-4f16-bf7f-25c4f6fe8a11-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72368">build#72368</a></summary>


| test_class            | test_method                        | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_migrated_topic_data_integrity | {"include_groups": false, "params": {"cancellation": {"dir": "in", "stage": "executed"}, "use_alias": false}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72368#019953b6-0892-4a9c-84ea-d639aee9bd38 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72508">build#72508</a></summary>


| test_class              | test_method            | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:------------------------|:-----------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SchemaEvolutionE2ETests | test_old_schema_writer | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "promote_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72508#01995a08-d78f-42ff-8bca-1a79e92be8ce | FLAKY         | 20/21    | upstream reliability is '99.45799457994579'. current run reliability is '95.23809523809523'. drift is 4.2199 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72538">build#72538</a></summary>


| test_class            | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                            |
|:----------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_recovery_mode_rebalance_finish | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72538#01995d00-920d-48c4-be97-98bfad007ccc | FLAKY         | 18/21    | upstream reliability is '81.6260162601626'. current run reliability is '85.71428571428571'. drift is -4.08827 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_recovery_mode_rebalance_finish |
</details>
<!--/body-->

WillemKauf committed:
`ct/l1`: move `compaction::sink` to `compaction::in_memory_sink`

This was a rough swing at a smoke test with the IO implementations
which uses an in-memory buffer to filter and create an L1 object.

While helpful for testing, the `compaction::sink` will be evolving
far past this point. Copy the current implementation to
`tests/in_memory_sink.{cc/h}`. `compaction/sink.{cc/h}` will be changed
in a future commit soon after.

WillemKauf committed:
`ct/l1`: `compaction::sink` forward progress

Still very much an unfinished state, but switches the `compaction::sink`
to instead operate on a `staging_file` instead of an in-memory buffer,
with committed objects going to the `compaction_committer`.

WillemKauf committed:
`ct/l1`: `compaction::source` forward progress

Also a very rough state, but adding in use of the `L1` reader,
as well as a lot of required state within the `compaction_source`.

WillemKauf committed:
`ct/l1`: move `l1_reader_test` to `l1_reader_fixture`

I'll want to use this elsewhere in a future commit.

WillemKauf committed:
`ct/l1/compaction`: add `Reducer` test with new definitions

This test proves that a compaction `sink`, `source`, `reducer`, `metastore`, and
`committer` can all work in tandem to produce updates to L1 data committed
to the `metastore`.

rockwotj commented on src/v/cloud_topics/level_one/compaction/sink.cc:
```diff
@@ -12,63 +12,143 @@
     co_return;
 }
 
+void compaction_sink::reset_metadata() {
+    _base_offset = kafka::offset::min();
+    _last_offset = kafka::offset::min();
+    _max_timestamp = model::timestamp::min();
+}
+
+void compaction_sink::update_metadata(const model::record_batch& b) {
+    if (_base_offset == kafka::offset::min()) {
+        _base_offset = model::offset_cast(b.base_offset());
+    }
+
+    _last_offset = model::offset_cast(b.last_offset());
+    _max_timestamp = std::max(_max_timestamp, b.header().max_timestamp);
+}
```
The builder tracks all this so you be able to pluck it out of the returned footer: https://github.com/redpanda-data/redpanda/blob/b9b51db9dc171f6c877691d6d257838325d29912/src/v/cloud_topics/level_one/common/object.h#L118-L123

WillemKauf replied:
Oh duh yeah. Thanks, removed

rockwotj commented on src/v/cloud_topics/level_one/compaction/source.cc:
```diff
@@ -10,21 +10,211 @@
+      std::make_unique<level_one_log_reader_impl>(
+        config, _ntp, _tp, _metastore, _io));
+
+    auto stats = co_await rdr.consume(
+      compaction_filter{sink, *_map, _ntp}, model::no_timeout);
```
I feel like we'll need to plumb in the abort source here eventually

WillemKauf replied:
Agree, I was surprised the `L1` reader doesn't use one?

WillemKauf replied:
The `abort_source` is in the `cloud_topic_log_reader_config`, so I think it is considered plumbed in here?

rockwotj replied:
Ah right 🙃

rockwotj reviewed (commented):
Very nice! I need to get more familiar with the code in `src/v/compaction` but this looks nice. Only one comment about maintaining builder metadata in two places, but if you're going to need that state later before finishing the building then it makes sense to duplicate.

dotnwat reviewed (commented):
lgtm

andrwng commented on src/v/cloud_topics/level_one/compaction/source.h:
```diff
@@ -10,18 +10,52 @@
+    bool preempted() const;
+
+private:
+    model::ntp _ntp;
+    model::topic_id_partition _tp;
+    offset_interval_set _dirty_ranges;
+    offset_interval_set _removable_tombstone_ranges;
```
nit: can these be const? If anything so it's harder to confuse them for some of the other members?

andrwng commented on src/v/cloud_topics/level_one/compaction/source.h:
```diff
@@ -10,18 +10,52 @@
+    ss::abort_source& _as;
+    compaction_job_state& _state;
+
+    using interval_vec = chunked_vector<offset_interval_set::interval>;
+    interval_vec _dirty_range_intervals;
+    interval_vec::const_iterator _map_building_it;
+
+    offset_interval_set _indexed_intervals;
```
nit: would appreciate some light comments here

andrwng commented on src/v/cloud_topics/level_one/compaction/tests/reducer_test.cc:
```diff
@@ -66,3 +83,129 @@ TEST(ReducerTest, Batches) {
+class ReducerTestFixture : public l1::l1_reader_fixture {
+public:
+    l1::compaction_committer::updates_t
+    steal_updates_from_committer(l1::compaction_committer& committer) {
```
nit: more commonly "take_..."

andrwng reviewed (commented):
👍
